### ISSUE TITLE
Unable to delete files on Fedora 28 (gvfs-trash is deprecated)

### ISSUE THREAD

**Issue by Devligue** [atom/atom#17452]
### Prerequisites

* [x] Put an X between the brackets on this line if you have done all of the following:
    * Reproduced the problem in Safe Mode: https://flight-manual.atom.io/hacking-atom/sections/debugging/#using-safe-mode
    * Followed all applicable steps in the debugging guide: https://flight-manual.atom.io/hacking-atom/sections/debugging/
    * Checked the FAQs on the message board for common solutions: https://discuss.atom.io/c/faq
    * Checked that your issue isn't already filed: https://github.com/issues?utf8=✓&q=is%3Aissue+user%3Aatom
    * Checked that there is not already an Atom package that provides the described functionality: https://atom.io/packages

### Description

My issue is the same as in #15949 except it is not fixed. 

I am unable to delete file (or directory). When trying, the following message appears: `Is gvfs-trash installed?`

### Steps to Reproduce

1. Try to delete any file or directory
2. ???
3. Profit

**Expected behavior:** Deleting file should delete file

**Actual behavior:** The file is not deleted

**Reproduces how often:** Every time

### Versions

```
Atom    : 1.27.1
Electron: 1.7.15
Chrome  : 58.0.3029.110
Node    : 7.9.0
```
```
apm  1.19.0
npm  3.10.10
node 6.9.5 x64
atom 1.27.1
python 2.7.15
git 2.17.1
```
```
OS: Fedora 28
```

### Additional Information
```
$ gvfs-trash
This tool has been deprecated, use 'gio trash' instead.
See 'gio help trash' for more info.
```
The `gio trash` is supposedly implemented since electron 1.7.2 and for some folks out there this problem was fixed with `Atom 1.25` (which included upgrade to `Electron 1.7.11`) but apparently I am on even newer Atom version, with even newer Electron and it still happens.

**Comment by rsese** [comment]
Thanks for opening a new issue - is `gio` available on Fedora 28?

**Comment by Devligue** [comment]
Yes, `gio` is available and installed by default

**Comment by rsese** [comment]
Hmmm, not sure why it's not working then - any difference if you specifically set an `ELECTRON_TRASH` environment variable to `gio` and then restart Atom?

**Comment by Devligue** [comment]
It changes nothing :thinking:

**Comment by rsese** [comment]
And if you just use `gio trash` from the command line it works fine?  E.g.:

```
echo hello > delete-me.txt
gio trash delete-me.txt
```

Also, are you getting the exact same error message you mentioned ("When trying, the following message appears: Is gvfs-trash installed?") after you set the `ELECTRON_TRASH` environment variable to `gio`?

**Comment by Devligue** [comment]
The `echo hello > delete-me.txt; gio trash delete-me.txt` works just fine normally but crashes on the ntfs partition. As I see this, it is not an Atom bug, however the error message it kept returning me was quite misleading.

I didn't solve the problem yet, but I am closing this since it is not related to Atom. Sorry for trouble!

**Comment by 50Wliu** [comment]
> however the error message it kept returning me was quite misleading.

Yeah, that's the default error message.  Unfortunately, Electron's moveToTrash method doesn't say _why_ it failed, just that it did.  We've found that the most common reason is that `gvfs-trash` isn't installed, which is why we decided to include it in the error message.

**Comment by cutephoton** [comment]
@Devligue This is a legit bug. The _tldr answer_ is that `_g_file_trash` may not be supported. In order for `_g_file_trash` to succeed, the following conditions must be met:

- The file being moved to trash is on the same partition as your home folder
**OR**
- A trash folder already exists or can be created at the mount point and,
- The trash folder is considered 'safe/secure' by verifying UID and restrictive permissions

**How this should probably be fixed...**

Atom does not have a fallback mechanism like offering the ability to permanently remove a file instead. This would address bring atom in line with gnome's graceful behaviour. The attached image show how  Gnome Files (aka Nautilus) prompt the user to permanently delete the file (when trash not supported)

![gnome's solution](https://user-images.githubusercontent.com/19512121/43511743-0dac32c0-952e-11e8-8af8-e0ef6cb916d0.png)

**Impact: Users with multiple disks and partitions**

Fedora/Redhat/etc: _Impacted_ due to default partition scheme that separates `/` and `/home` in to partitions. Files outside of the `/home` partition cannot be moved to trash.

Ubuntu: Less likely due to partition layout (`/home` is part of the `/` partition)

Conditions where users will be impacted:
- File systems with unix permissions:
  - _Impacted with workaround_ due to typical restrictive top-level directory permissions (root)
- FAT/non-unix file systems:
  - User Session Mount (i.e. `/run/media/$USER/disk-label`): _No Impact_
  - Fstab: _Impacted_ even with permissive umask. The default uid/gid is root. It will be unable to satisfy the requirements of a trash folder without additional options (uid/gid/umask).
- Network shared folders: _Unknown/Did not test._

**Notes/Testing the Root Cause**

Note: Though I am confident that my analysis is decent enough, much of the code I was referencing was unfamiliar/new to me.

Initially I encountered this issue when I put files in a certain location like the reporter above. I created a delete-me file test as suggested above in the relevant directory, `/opt/cupenv`. I ran the command  `strace gio trash delete-me.txt`. A abridged version of the output is here:

```
lstat("/opt/cupenv/delete-me.txt", {st_mode=S_IFREG|0664, st_size=6, ...}) = 0
...
lstat("/.Trash-1000", 0x7ffde5ffb5b0)   = -1 ENOENT (No such file or directory)
mkdir("/.Trash-1000", 0700)             = -1 EACCES (Permission denied)
```
That's unexpected. The API documentation for [g_file_trash](https://developer.gnome.org/gio/stable/GFile.html#g-file-trash) lacks some level of specificity.

> Sends file to the "Trashcan", if possible. This is similar to deleting it, but the user can recover it before emptying the trashcan. Not all file systems support trashing, so this call can return the G_IO_ERROR_NOT_SUPPORTED error. 

One might assume (as I did): Given a path, if the user has permission to modify/delete the file, then GIO's g_file_trash API should succeed at removing the file in some manner. Perhaps, if trash functionality is not available, then there might be a fallback mechanism. In the case of `gio trash` I expected there to be a -f force option that would prioritize trash over permanently deleting the file. (`gio trash -f` only ignores files that don't exist)

Glib appears to implement GIO local file access using [glocalfile.c](https://github.com/GNOME/glib/blob/83a4cab12c2d00dbfe6013d071cff2da310109a4/gio/glocalfile.c#L1899). The trash algorithm looks like this:

- Is the path on the same partition as the user's home directory? If so, move the file to the home trash folder if possible and exit.
- Given the path, locate the mount point top level directory (denoted as `$topdir` in source)
  - In my case, `/opt/cupenv/delete_me.txt` -> `/` (the primary partition)
  - A more typical case: `/run/media/$USER/disk-label/a/b/c/d` would be `/run/media/$USER/disk-label`
- If `$topdir` is found, pick one of `$topdir/.Trash/$UID` or `$topdir/.Trash-UID` folder.
  - If `$topdir/.Trash-UID` and `$topdir/.Trash/$UID` exists: Validate proper UID and file permission (or fail)
  - Try to create folder `$topdir/.Trash-$UID`. Validate proper UID and file permission (or fail)

At the end of the function we find [`G_IO_ERROR_NOT_SUPPORTED` is returned](https://github.com/GNOME/glib/blob/83a4cab12c2d00dbfe6013d071cff2da310109a4/gio/glocalfile.c#L2059) when such a folder is not found and cannot be created.

```
if (trashdir == NULL)
	{
	  g_free (topdir);
          g_set_io_error (error,
                          _("Unable to find or create trash directory for %s"),
                          file, G_IO_ERROR_NOT_SUPPORTED);
	  return FALSE;
}
```

I confirmed my understanding by creating a top level trash folder and using gio trash command.

```
sudo mkdir /.Trash-1000
sudo chmod 0700 /.Trash-1000
sudo chown fosterb:fosterb /.Trash-1000
cd /opt/cupenv && echo hello > delete-me.txt; gio trash delete-me.txt
```
Files located on the same partition as your home directory can always be moved to your user trash folder. On my operating system, the paths `/` and `/home` are separate partitions. I created a file `/home/tmp/x` where `tmp` and `x` are owned by my user. In this case, the files get moved to your `$HOME/.local/share/Trash` folder.

```
lstat("/home/tmp/x", {st_mode=S_IFREG|0644, st_size=0, ...}) = 0
stat("/home/fosterb", {st_mode=S_IFDIR|0700, st_size=4096, ...}) = 0
access("/home", F_OK)                   = 0
....
stat("/home/fosterb/.local/share", {st_mode=S_IFDIR|0755, st_size=4096, ...}) = 0
access("/home/fosterb/.local/share/Trash", F_OK) = 0
stat("/home/fosterb/.local/share/Trash", {st_mode=S_IFDIR|0700, st_size=4096, ...}) = 0
mkdir("/home/fosterb/.local/share/Trash/info", 0700) = -1 EEXIST (File exists)
mkdir("/home/fosterb/.local/share/Trash/files", 0700) = -1 EEXIST (File exists)
openat(AT_FDCWD, "/home/fosterb/.local/share/Trash/info/x.trashinfo", O_RDONLY|O_CREAT|O_EXCL, 0666) = 7
```
But when mounting disks (external or internal) using fstab or mount commands (as opposed to session-based mounting) issues can arise. I tested a FAT file system with umask=0000 and uid/gid set to root.

```
lstat("/mnt/giotest", {st_mode=S_IFDIR|0777, st_size=8192, ...}) = 0
lstat("/mnt", {st_mode=S_IFDIR|0755, st_size=4096, ...}) = 0
lstat("/mnt/giotest/.Trash", 0x7fffec5502c0) = -1 ENOENT (No such file or directory)
lstat("/mnt/giotest/.Trash-1000", 0x7fffec550230) = -1 ENOENT (No such file or directory)
mkdir("/mnt/giotest/.Trash-1000", 0700) = 0
lstat("/mnt/giotest/.Trash-1000", {st_mode=S_IFDIR|0777, st_size=8192, ...}) = 0
unlink("/mnt/giotest/.Trash-1000")      = -1 EISDIR (Is a directory)
rmdir("/mnt/giotest/.Trash-1000")       = 0
file:///mnt/giotest/x: Unable to find or create trash directory for /mnt/giotest/x
```

You can see gio trash going through the motions. The operations success but the command still reports an error.
```
[fosterb@rose giotest]$ ls -lah
total 12K
drwxrwxrwx. 2 root root 8.0K Aug  1 00:36 .
drwxr-xr-x. 4 root root 4.0K Aug  1 00:30 ..
-rwxrwxrwx. 1 root root    0 Aug  1 01:23 x
```

GLib/Gio will not write to a trash directory with incorrect permissions/ownership. (Security?) Without unix permissions, trash will always fail on these mount points. Modifying fstab to appear similar to options used during session mounts (UID/GID set to user vs root, umask is set appropriately). Conveniently the [source code](https://github.com/GNOME/glib/blob/83a4cab12c2d00dbfe6013d071cff2da310109a4/gio/glocalfile.c#L2046) has a comment that seems to strongly imply this is a known/expected.

```
		  /* Ensure that the created dir has the right uid etc.
		     This might fail on e.g. a FAT dir */
```
Most of this was unnecessary, but I figured I'd show my work.

**Comment by yodatak** [comment]
I got the same problem on feodra 29 on a fresh install can we reopen this bug ?

**Comment by cutephoton** [comment]
@Devligue can reopen this

**Comment by Devligue** [comment]
I may have no additional information about the issue, but I am sure @cutephoton provided enough insight. There are also other people who have similar unsolved problems. I am reopening this.

**Comment by cutephoton** [comment]
I reproduced on the latest version still. **shrug** I don't see how flatpak and rpm/deb would differ. It only fails when the preconditions listed above are met -- a partition where you don't have permission to create a trash folder under your user ID.

**Comment by yodatak** [comment]
With flatpak i got no error but it don't get delete sorry about this misinformation. Can it be a selinux policy problem ?

**Comment by cutephoton** [comment]
No. Not selinux. It has to do with how trash folders work. My rather verbose explanation gets in to the nitty gritty for where atom's behavior is deficient.  The short answer is that atom only asks to move files to trash and does not delete files. If a trash folder cannot be created (multiple causes all related to permissions) than it fails. Atom should inquire with user if they should permanently delete the files instead of failing.

**Comment by mkungla** [comment]
As @yodatak also mentioned that Fedora 29 fresh install has this problem since `gvfs-trash` has finally after many deprecation notices been removed  and replaced with  `gio`. So as @rsese mentioned you need to set `ELECTRON_TRASH=gio` e.g. add it to your `~/.bash_profile`. 

logout and login or `source ~/.bash_profile` and trashing will work

**Comment by cutephoton** [comment]
Gio operates the same way. I used the Gio command line to manually step through this...

**Comment by mh-cbon** [comment]
running fedora 29, i have installed atom from packages, since recently I faced the same error as described here.

```
sudo dnf list installed atom
Paquets installés
atom.x86_64                                                1.33.0-0.1                                                 @Atom
```

I think this bug deserves an action in order to improve the package installation, or the runtime, to detect gio/gvfs and make this work out of the box.

This really is an ugly bug for the end user experience.

**Comment by cristianlivella** [comment]
Is there any workaround to delete files with Atom on Fedora 29?

**Comment by edlefebvre** [comment]
> So as @rsese mentioned you need to set `ELECTRON_TRASH=gio` e.g. add it to your `~/.bash_profile`.
> 
> logout and login or `source ~/.bash_profile` and trashing will work

This is not working on fedora 29, even with gio installed and ELECTRON_TRASH variable set to gio.

**Comment by cutephoton** [comment]
There are no workarounds.

**Comment by olegjs** [comment]
> > So as @rsese mentioned you need to set `ELECTRON_TRASH=gio` e.g. add it to your `~/.bash_profile`.
> > logout and login or `source ~/.bash_profile` and trashing will work
> 
> This is not working on fedora 29, even with gio installed and ELECTRON_TRASH variable set to gio.

Works on my Fedora 29 -- had to restart terminal and restart atom from that terminal.

**Comment by cutephoton** [comment]
That is a not a fix. There are preconditions as documented above to reproduce the bug.

**Comment by shivdhar** [comment]
Would love for this to be fixed (on my Fedora 29 installed via the official RPM).

**Comment by histamineblkr** [comment]
This bug is also affecting me on a fresh install of Fedora 29 and Atom 1.33.1.

If I try to delete a file within atom, I get the following error:
![gvfs-gio-error](https://user-images.githubusercontent.com/5628330/50742193-8be3d980-11bc-11e9-9d98-8c307d563839.png)

If I use gio on the command line:
```
$ gio trash /home/bauthier/Tools/ansible_local/roles/work_packages/files/test.txt
$
```
no issue and it deletes as expected.

My system partitions are broken up like so:
```
$ df -h
Filesystem               Size  Used Avail Use% Mounted on
devtmpfs                 7.8G     0  7.8G   0% /dev
tmpfs                    7.8G  225M  7.6G   3% /dev/shm
tmpfs                    7.8G  2.0M  7.8G   1% /run
tmpfs                    7.8G     0  7.8G   0% /sys/fs/cgroup
/dev/mapper/fedora-root   49G   11G   36G  24% /
tmpfs                    7.8G  236K  7.8G   1% /tmp
/dev/sda2                976M  160M  750M  18% /boot
/dev/mapper/fedora-home  414G  892M  392G   1% /home
/dev/sda1               1022M   21M 1002M   3% /boot/efi
tmpfs                    1.6G  8.4M  1.6G   1% /run/user/1000
```
If I explicitly set an environment variable like this within my **~/.bashrc**:
```
export ELECTRON_TRASH=gio
```
it does delete within atom correctly.

Having to set this environment variable is less than ideal. I read through cutephoton's research, testing, and analysis and I believe mine works because **/home** and **/** are on different mounts, but the same partition. Perhaps LVM versus standard partitions?

Either way, I would imagine the handling of "gvfs" or "gio" would be done by atom and set correctly during start time of the application. I haven't read through the atom code so there may have been a reason to couple that functionality to "gvfs" rather than dynamically choosing a suitable trash program and application startup.

**Comment by yodatak** [comment]
https://github.com/electron/electron/issues/15011

**Comment by den-is** [comment]
same here with fedora 29 and atom 1.34.0

that workaround has helped me https://github.com/atom/atom/issues/15949#issuecomment-434231263

**Comment by abudawud** [comment]
same here with fedora 29 and atom 1.33.0x64
`~$ lsb_release -a`
`LSB Version:	:core-4.1-amd64:core-4.1-noarch
Distributor ID:	Fedora
Description:	Fedora release 29 (Twenty Nine)
Release:	29
Codename:	TwentyNine`

**Comment by Arcesilas** [comment]
@nelson6e65 It worked for me too, on Fedora 29, but not undefinitely: after a moment, Atom could not delete files anymore, just like the variable was no more defined. Had to restart atom from a terminal, which was not an acceptable solution. On the contrary, the workaround quoted by @den-is works perfectly.

**Comment by nelson6e65** [comment]
@Arcanemagus Hmm... I'm not experienced that, but I'll use it for a while to see what happens.

I would like to keep it simple and avoid "cheat" Atom that is using the "real" `gvfs-trash`. I don't know if usage changes or etc... _But especially if I'm not a user with no **`sudo` powers**_. 

So, I hope don't have to. Thanks for your info. I'll keep an eye on it.

**Comment by Arcesilas** [comment]
@nelson6e65 (you mistyped my nickname...)
Without sudo powers, you can place the script anywhere which is in your PATH (for example, `~/bin`).
On Fedora 29, the real `gfvs-trash` has been removed and is not meant to reappear, so you can have a script with that name without worrying about any conflict.

**Comment by rdlmda** [comment]
I think it's bad user experience having to experience the bug, then google it, find this thread, add a workaround, restart the terminal, restart the session, restart Atom...

**Comment by cutephoton** [comment]
Yes, as reported, the ELECTRON_TRASH environment variable is needed. This is the issue referenced #15949 bug. The issue reported here is not related to #15949. The error message related to gvfs-trash is a blunt tool (even access denied errors produce the same error message) so I understand why people are responding here.

The reporter clarifies one of the key symptoms here:
https://github.com/atom/atom/issues/17452#issuecomment-396407349

And that led to my deep dive. My deep dive details the specific cases where trash becomes unavailable (i.e. when a trash folder cannot be safely created). 
https://github.com/atom/atom/issues/17452#issuecomment-409518547

**I just revalidated my findings and unsurprisingly gio trash still fails.**

Anyhow, I wanted to clarify this so it doesn't get lost when/if atom devs look in to this issue.

**Comment by Caseraw** [comment]
> @histamineblkr  
>  
> If I explicitly set an environment variable like this within my **~/.bashrc**:
> ```shell
> export ELECTRON_TRASH=gio
> ```

This worked for me also!  
Thanks @histamineblkr

**Comment by TJProgrammer** [comment]
Same here:
```
atom -v
Atom    : 1.34.0
Electron: 2.0.16
Chrome  : 61.0.3163.100
Node    : 8.9.3

gio --version
2.58.3

Fedora 29
4.20.6-200.fc29.x86_64

```

**Comment by cutephoton** [comment]
Honest question: does anybody read what I've been posting about this issue? There is a bug here and posting that this is fixed is misleading. Maybe I should just submit new issue request specific to the limitations of the the "delete" functionality.

**Comment by james-portman** [comment]
If the above fixes don't work, try this (worked on Fedora 29)
add: `export ELECTRON_TRASH=gio` to the end of your `~/.bashrc`.

You will need to open a new terminal or log out/in

**Comment by Arcesilas** [comment]
@james-portman Did you actually read the thread?

**Comment by james-portman** [comment]
I have now..
I got as far as the first person saying to try .bash_profile, oh well

**Comment by mkungla** [comment]
[Electron](https://github.com/electron/electron) introduced the fix already in October 2018 

https://github.com/electron/electron/commit/bea4de926220c5b24d51511d7ef066e33f1ff6d4#diff-6618f017d105c6bd855015f12c212103R19

https://github.com/electron/electron/blob/bea4de926220c5b24d51511d7ef066e33f1ff6d4/atom/common/platform_util_linux.cc#L17...L20

How difficult is it to get it fixed?

**Comment by cutephoton** [comment]
@james-portman  @mkungla This is a different issue as I have noted a couple times. Unfortunately, this unbelievably dumb error message has caused lots of confusion.

**Comment by james-portman** [comment]
There was quite a long thread so myself and others are skipping through,
would be helpful to change the original post to point to your comments
@cutephoton

**Comment by Devligue** [comment]
I added relevant @cutephoton comments to the original question.

**Comment by lee-dohm** [comment]
If the required solution here is that Atom implement the ability to permanently delete files, that isn't something the maintainer team is going to do. We've had the ability to permanently delete files in the past and it created too many support headaches when people accidentally permanently deleted files (even with appropriate warnings) and blamed Atom. If someone wants the ability to permanently delete files, then it can be implemented in a community package.

If I'm misunderstanding what is being requested here, please let me know.

**Comment by Arcesilas** [comment]
I guess the solution is to use gio when gvfs-trash is not present, because it's been deprecated for quite some time.
Part of the solution would be to make the error message relevant.
Some workarounds have been provided and they work. So, the solution could consider them.

**Comment by lee-dohm** [comment]
It explicitly states in the updated comments added to the first message:

> Atom does not have a fallback mechanism like offering the ability to permanently remove a file instead.

Other than that, I didn't see a proposed solution. So that's why I asked.

**Comment by cutephoton** [comment]
#response_container_BBPPID{font-family: initial; font-size:initial; color: initial;} This is the standard approach used by most software when confronted with the "trash unavailability" problem. If you want to be absolutely sure, put a delay on the dialog or a check box "I really want to delete this file and understand the risks."At the very least, fix the error message.                                                                                                                                                                        Sent from my BlackBerry - the most secure mobile device                                                                                                                                                                           From: notifications@github.comSent: March 13, 2019 10:07 AMTo: atom@noreply.github.comReply-to: reply@reply.github.comCc: fosterb@edgeandvertex.org; mention@noreply.github.comSubject: Re: [atom/atom] Unable to delete files on Fedora 28 (gvfs-trash is deprecated) (#17452)  It explicitly states in the updated comments added to the first message:

Atom does not have a fallback mechanism like offering the ability to permanently remove a file instead.

Other than that, I didn't see a proposed solution. So that's why I asked.

**Comment by Arcesilas** [comment]
Yes, *one* solution could be to have a fallback mechanism, *like* offering the ability to permanently remove a file. You say it's not gonna happen, very well. It was just a suggestion.

Electron does support gio, Atom does not. *Another* solution could be to support gio, since gvfs-trash has been deprecated for quite a long time on Fedora (for example).

One possible solution would be to have a configuration option to define the delete command. By default it could be `gvfs-trash`, and user could define it to `gio trash`. But it **requires** the error message to be explicit.

You assumed the most common error was because `gvfs-trash` was not installed. This thread is the proof it's no more the case. Please, update.

**Comment by cutephoton** [comment]
#response_container_BBPPID{font-family: initial; font-size:initial; color: initial;}  I'm confused by this comment. Gvfs-trash or Gio trash both do the same thing and both behave the same way. That's not material to this issue. (I think the confusion goes back to the fact we're dealing with two separate issues here. Fixing one doesn't fix the other. I think we are in agreement, but not totally sure.)(Aside: Why aren't they just using native APIs? It's crazy to me that they are using command line tools to implement this.)                                                                                                                                                                        Sent from my BlackBerry - the most secure mobile device                                                                                                                                                                           From: notifications@github.comSent: March 13, 2019 10:18 AMTo: atom@noreply.github.comReply-to: reply@reply.github.comCc: fosterb@edgeandvertex.org; mention@noreply.github.comSubject: Re: [atom/atom] Unable to delete files on Fedora 28 (gvfs-trash is deprecated) (#17452)  Yes, one solution could be to have a fallback mechanism, like offering the ability to permanently remove a file. You say it's not gonna happen, very well. It was just a suggestion.
Electron does support gio, Atom does not. Another solution could be to support gio, since gvfs-trash has been deprecated for quite a long time on Fedora (for example).
One possible solution would be to have a configuration option to define the delete command. By default it could be gvfs-trash, and user could define it to gio trash. But it requires the error message to be explicit.
You assumed the most common error was because gvfs-trash was not installed. This thread is the proof it's no more the case. Please, update.

**Comment by rdazvd** [comment]
> If the above fixes don't work (`~/.bash_profile`), try this (worked on Fedora 29)
> add: `export ELECTRON_TRASH=gio` to the end of your `~/.bashrc`.
> 
> You will need to open a new terminal or log out/in

This solution works perfectly on Fedora 28/29. In strict terms, I would argue that this is not a bug intrinsically related to Atom (i.e., this is not an error in the source) ─ if much, the "is gvfs-trash installed" error message is misleading to Fedora users and perhaps could be changed into something more generic ─ but given that there is a workaround shouldn't this issue be closed?

**Comment by cutephoton** [comment]
#response_container_BBPPID{font-family: initial; font-size:initial; color: initial;} As I have repeatedly mentioned, this bug is distinct from the solution you are copy and pasting from.                                 Sent from my BlackBerry - the most secure mobile device                                                                                                                                                                           From: notifications@github.comSent: April 19, 2019 12:45 PMTo: atom@noreply.github.comReply-to: reply@reply.github.comCc: fosterb@edgeandvertex.org; mention@noreply.github.comSubject: Re: [atom/atom] Unable to delete files on Fedora 28 (gvfs-trash is deprecated) (#17452)  

If the above fixes don't work (~/.bash_profile), try this (worked on Fedora 29)

add: export ELECTRON_TRASH=gio to the end of your ~/.bashrc.

You will need to open a new terminal or log out/in



This solution works perfectly on Fedora 28/29. In strict terms, I would argue that this is not a bug intrinsically related to Atom (i.e., this is not an error in the source) ─ if much, the "is gvfs-trash installed" error message is misleading to Fedora users and perhaps could be changed into something more generic ─ but given that there is a workaround shouldn't this issue be closed?

**Comment by rugk** [comment]
BGTW for anyone on Fedora who wants an easy "fix": The flatpak version of Atom works fine… :smiley:

**Comment by rdazvd** [comment]
> BTW for anyone on Fedora who wants an easy "fix": The flatpak version of Atom works fine…

Installing the flatpak version worked for me as well (Fedora 29). It is a solution, although installing via flatpak means giving up CLI tools.

**Comment by cutephoton** [comment]
I would direct people to actually read the bug report before commenting. Somebody please lock down this issue.

**Comment by rugk** [comment]
> although installing via flatpak means giving up CLI tools.

You _can_ expose them, so they spawn outside of the flatpak. See https://github.com/rugk/atom-flatpak-wrapper/.

Anyway, this is getting off-topic.

@cutephoton See no reason for a lock here. The issue is not really fixed or so… or Fedora has to fix it (then one needs to notify them), whatever…

**Comment by cldtech** [comment]
A small trick that is not a solution but probably do it for most. Install the permanent-delete plugin and use shift+del. It is made to bypass the trash can but it's a quick work around. 〰️
[](https://atom.io/packages/permanent-delete)

**Comment by Arcesilas** [comment]
> A small trick that is not a solution but probably do it for most. Install the permanent-delete plugin and use shift+del. It is made to bypass the trash can but it's a quick work around. wavy_dash

There are already better workarounds which do not permanently delete files.

**Comment by cldtech** [comment]
If what you have to say don't bring something to the conversation, you are
just polluting the web, who are you to decide if someone's gonna need this
information or not.

**Comment by cldtech** [comment]
> > A small trick that is not a solution but probably do it for most. Install the permanent-delete plugin and use shift+del. It is made to bypass the trash can but it's a quick work around. wavy_dash
> 
> There are already better workarounds which do not permanently delete files.

If what you have to say don't bring something to the conversation, you are
just polluting the web, who are you to decide if someone's gonna need this
information or not. This is an option and it's related to the subject, that's all.

**Comment by Arcesilas** [comment]
Let's continue polluting, then: permanently deleting files is NOT a workaround to sending to trash. That's all.

**Comment by cldtech** [comment]
https://dictionary.cambridge.org/us/dictionary/english/workaround

**Comment by cldtech** [comment]
> > A small trick that is not a solution but probably do it for most. Install the permanent-delete plugin and use shift+del. It is made to bypass the trash can but it's a quick work around. wavy_dash
> 
> There are already better workarounds which do not permanently delete files.

You just said there is BETTER workaround. Decide yourself.

**Comment by cldtech** [comment]
> Let's continue polluting, then: permanently deleting files is NOT a workaround to sending to trash. That's all.

The original question say; I am unable to delete file (or directory). NOT TRASH.
Now get over yourself your not god.

**Comment by Arcesilas** [comment]
I love you too.

**Comment by cldtech** [comment]
> I love you too.

I'm glad we found a workaround ♥

**Comment by Arcesilas** [comment]
Atom DOES NOT permanently delete files. The initial problem, as you mention it, is that files are not deleted. Actually, the problem is that "files are not sent to trash", since Atom does not provide any alternative when it cannot send to trash. But what Atom does is ALWAYS to send to trash. Therefore, permanently deleting files is NOT a workaround, according to the definition you mentioned. Your "solution" actually allows user to remove files. But "removing" files is a truncated vision of the problem. The error message is a consequence of the problem. Your "solution" allows to remove a file and not have the error message anymore. It does not allow to recover deleted file. A workaround should allow to recover the file.

Examples of real workarounds (allow to remove files, error message does not show anymore and files are recoverable):
- https://github.com/atom/atom/issues/17452#issuecomment-457959922
- https://github.com/atom/atom/issues/15949#issuecomment-434231263

**Comment by cldtech** [comment]
> Your "solution" actually allows user to remove files.

You're kidding, have you even read my comment; 

> A small trick that is not a solution but probably do it for most.

**Comment by Arcesilas** [comment]
And I was polluting the web, while you're discussing each word we write... Are we really going to have a discussion about grammar and linguistics?

**Comment by cldtech** [comment]
What you mean is semantic and terminology and it;s another example of your lack of precision that lead you here. But yes, i'll always have time for weeds, if you don't get rid of it it will spread across the garden.

**Comment by Arcesilas** [comment]
You're very funny... (j/k)

**Comment by cldtech** [comment]
I specifically wrote that this is not a solution only to avoid your kind of insecure almost programmer. Start coding for real without Lego style "key in hand" stuff like WordPress that i'm sure your using and come back when you're fully formed.

**Comment by Arcesilas** [comment]
Thanks for your personnal attacks which are very constructive and absolutely not polluting the thread.

**Comment by cldtech** [comment]
We are two here! i take 50% gladly and remember you i;m not the one you started this! You're just tired of getting you're errors highlighted.

**Comment by Arcesilas** [comment]
Ok. Be happy with your superiority.

**Comment by cldtech** [comment]
You're not gonna have your beloved last words.

**Comment by lee-dohm** [comment]
@cldtech and @Arcesilas your comments were deleted as a violation of the [Atom Code of Conduct](https://github.com/atom/atom/blob/master/CODE_OF_CONDUCT.md) as they were insulting or derogatory.

**Comment by rafeca** [comment]
Atom is using [an electron API](https://electronjs.org/docs/api/shell#shellmoveitemtotrashfullpath) to move items to the trash, this API used to default to `gvfs-trash` but this got changed in electron v3 to default to `gio`.

We're planning to upgrade to Electron v3 very soon ([WIP PR](https://github.com/atom/atom/pull/18916)), so this issue will get resolved quite soon.
